c++/src/capnp/any.c++

```diff
@@ -77,4 +79,127 @@ kj::Own<ClientHook> AnyPointer::Pipeline::asCap() {
+
+  if(dataSizeL != dataSizeR) {
+    return StructEqualityResult::NOT_EQUAL;
+  }
+
+  if(0 != memcmp(dataL.begin(), dataR.begin(), dataSizeL * sizeof(word))) {
+    return StructEqualityResult::NOT_EQUAL;
+  }
+
+  auto ptrsL = left.getPointerSection();
+  auto ptrsR = right.getPointerSection();
+
+  size_t i = 0;
+
+  for(; i < kj::min(ptrsL.size(), ptrsR.size()); i++) {
+    auto l = ptrsL[i];
+    auto r = ptrsR[i];
+    switch(equal(l, r)) {
+    case StructEqualityResult::EQUAL:
+      break;
+    case StructEqualityResult::NOT_EQUAL:
+      return StructEqualityResult::NOT_EQUAL;
+    case StructEqualityResult::UNKNOWN_CONTAINS_CAPS:
+      return StructEqualityResult::UNKNOWN_CONTAINS_CAPS;
+    }
+  }
+
+  return StructEqualityResult::EQUAL;
+
+}
+
+StructEqualityResult equal(AnyList::Reader left, AnyList::Reader right) {
+  if(left.size() != right.size()) {
+    return StructEqualityResult::NOT_EQUAL;
+  }
+  switch(left.getElementSize()) {
+  case ElementSize::VOID:
+  case ElementSize::BIT:
+  case ElementSize::BYTE:
+  case ElementSize::TWO_BYTES:
+  case ElementSize::FOUR_BYTES:
+  case ElementSize::EIGHT_BYTES:
+    if(left.getElementSize() == right.getElementSize()) {
+      if(memcmp(left.getData().begin(), right.getData().begin(), left.getData().size()) == 0) {
+        return StructEqualityResult::EQUAL;
+      } else {
+        return StructEqualityResult::NOT_EQUAL;
+      }
+    } else {
+      return StructEqualityResult::NOT_EQUAL;
+    }
+  case ElementSize::POINTER:
+  case ElementSize::INLINE_COMPOSITE: {
+    auto llist = left.as<List<AnyStruct>>();
+    auto rlist = right.as<List<AnyStruct>>();
+    for(size_t i = 0; i < left.size(); i++) {
+      switch(equal(llist[i], rlist[i])) {
+      case StructEqualityResult::EQUAL:
+        break;
+      case StructEqualityResult::NOT_EQUAL:
+        return StructEqualityResult::NOT_EQUAL;
+      case StructEqualityResult::UNKNOWN_CONTAINS_CAPS:
+        return StructEqualityResult::UNKNOWN_CONTAINS_CAPS;
+      }
+    }
+    return StructEqualityResult::EQUAL;
+  }
+  }
+}
+
+StructEqualityResult equal(AnyPointer::Reader left, AnyPointer::Reader right) {
+  if(right.isCapability()) {
+    return StructEqualityResult::UNKNOWN_CONTAINS_CAPS;
+  }
+  if(left.isNull()) {
+    if(right.isNull()) {
+      return StructEqualityResult::EQUAL;
+    } else {
+      return StructEqualityResult::NOT_EQUAL;
+    }
+  } else if(left.isStruct()) {
+    if(right.isStruct()) {
+      return equal(left.getAs<AnyStruct>(), right.getAs<AnyStruct>());
+    } else {
+      return StructEqualityResult::NOT_EQUAL;
+    }
+  } else if(left.isList()) {
+    if(right.isList()) {
+      return equal(left.getAs<AnyList>(), right.getAs<AnyList>());
+    } else {
+      return StructEqualityResult::NOT_EQUAL;
+    }
+  } else if(left.isCapability()) {
+    return StructEqualityResult::UNKNOWN_CONTAINS_CAPS;
+  } else {
+    // There aren't currently any other types of pointers
+    KJ_FAIL_REQUIRE();
```

Comment:
Use KJ_UNREACHABLE. (I commonly follow a switch() that handles all possible enums with KJ_UNREACHABLE. If a new enumerant is added, the switch() will produce a compiler warning.)